python/ray/autoscaler/v2/tests/test_node_provider.py

```diff
@@ -1,97 +1,306 @@
-# coding: utf-8
+import logging
 import os
 import sys
-import unittest
+import time
+
+# coding: utf-8
+from collections import defaultdict
+from unittest.mock import MagicMock
 
 import pytest  # noqa
 
-from ray._private.test_utils import load_test_config
-from ray.autoscaler._private.event_summarizer import EventSummarizer
-from ray.autoscaler._private.node_launcher import BaseNodeLauncher
-from ray.autoscaler._private.node_provider_availability_tracker import (
-    NodeProviderAvailabilityTracker,
+import ray
+from ray._private.test_utils import get_test_config_path, wait_for_condition
+from ray.autoscaler._private.constants import (
+    AUTOSCALER_MAX_CONCURRENT_LAUNCHES,
+    AUTOSCALER_MAX_LAUNCH_BATCH,
+)
+from ray.autoscaler._private.fake_multi_node.node_provider import FakeMultiNodeProvider
+from ray.autoscaler.v2.instance_manager.config import FileConfigReader
+from ray.autoscaler.v2.instance_manager.node_provider import (
+    LaunchNodeError,
+    NodeProviderAdapter,
+    TerminateNodeError,
+    logger,
 )
-from ray.autoscaler.node_launch_exception import NodeLaunchException
-from ray.autoscaler.v2.instance_manager.config import AutoscalingConfig
-from ray.autoscaler.v2.instance_manager.node_provider import NodeProviderAdapter
-from ray.autoscaler.v2.tests.util import FakeCounter
-from ray.core.generated.instance_manager_pb2 import Instance
 from ray.tests.autoscaler_test_utils import MockProvider
 
+logger.setLevel(logging.DEBUG)
 
-class NodeProviderTest(unittest.TestCase):
-    def setUp(self):
-        self.base_provider = MockProvider()
-        self.availability_tracker = NodeProviderAvailabilityTracker()
-        self.node_launcher = BaseNodeLauncher(
-            self.base_provider,
-            FakeCounter(),
-            EventSummarizer(),
-            self.availability_tracker,
+
+class NodeProviderTestWrapper:
+    def __init__(
+        self,
+        kind: str,
+        max_concurrent_launches: int,
+        max_launch_batch_per_type: int,
+    ):
+        self.kind = kind
+
+        self.config_reader = FileConfigReader(
+            get_test_config_path("test_ray_complex.yaml"), skip_content_hash=True
+        )
+        config = self.config_reader.get_autoscaling_config()
+        self.ray_session = None
+
+        if self.kind == "mock":
```

Comment:
this is how I imagine we would add more node providers to the unit testing when we start supporting all of them (with mock client like aws client, or k8s client)